resip/stack/ConnectionBase.cxx

```diff
@@ -1030,6 +1030,16 @@ ConnectionBase::getCurrentWriteBuffer()
    return std::make_pair(mBuffer + mBufferPos, mBufferSize - mBufferPos);
 }
 
+
+
+// mike did this 03/30/18 -- this is a bug fix to the stack
+// note that thuis function was returning as a pointer the size of the buffer, 
+// which is not nesesarily the position after where the buffer has been filled by the first SSL_read() in TlsConnection::read()
+// Hence, the buffer was being increased in length whenever this function was invoked (even when unnnecessary), and the 
+// "extraBytes" were being placed in the wrong place within the buffer.
+// My version (I think) corrects that -- see below.
+
+/*
 char*
 ConnectionBase::getWriteBufferForExtraBytes(int extraBytes)
```

Comment:
>>So it seems this old code, used to assume that SSL_pending would only return a value > 0, if >>the >>SSL_read failed to read all the bytes because it filled the buffer provided to it.

Right.... the first SSL_read() always gets Connection::ChunkSize (or less) as the maximum number of bytes to read (see Connection::read(), line 339 ). When we consistently get messages with big SDPs, SSL_pending() always gets back >0.

>> If the SSL_read operation did fill the buffer, then I think the current code is fine. However, I'm >> guessing you've found cases where SSL_pending is returning a value that is greater than 0 and >> SSL_read didn't actually entirely fill the current buffer.

Well, on the first big message, that comes through on the connection, the first SSL_read() gets back "ChunkSize" bytes, and SSL_Pending() returns >0, and the buffer isn't big enough, and gets reallocated. On any subsequent messages, the first SSL_read() also gets back ChunkSize bytes, and then SSL_pending() again gets back >0..... but this time the buffer *is* probably big enough.

So, yeah, the old code kind of assumes that if extraBytes > 0, then it must be that the buffer is not big enough. But if you look at Connection::Read(), there is this:

size_t bytesToRead = resipMin(writePair.second,
                                 static_cast<size_t>(Connection::ChunkSize));

   resip_assert(bytesToRead > 0);

   int bytesRead = read(writePair.first, (int)bytesToRead);

So, bytesToRead is never going to be more than ChunkSize, so SSL_pending() will always return >0 on each new big message.... and with the old code the buffer gets re-allocated each time it gets called with extraBytes > 0.... and the pointer that gives back leads TlsConnection to put the extra bytes un-concatenated with the original ChunkSize-sized string of bytes read on the first SSL_read().

Reply:
Ah I see what's going on here.  For WSS we don't use up the buffer and reallocate each message, we just keep the same buffer around.  Note for WSS it seems that mConnState is never NewMessage.  So this fix needs to aware of the normal TCP connection use as well, as we need to make sure we don't break that.  :)

Reply:
I reviewed your original solution and I don't see how it make anything worse for TCP or TLS connections, in fact it removes entirely the assumption that the buffer is completely full before getWriteBufferForExtraBytes is called.

Reply:
I will plan on submitting this solution separate from this pull request, and I'll remove the pull request.  Thanks for all your efforts here, and I'm sorry for the delay.

Reply:
That's great. I'm glad I could be of some help.

Sorry, I never got a chance to remove those goofy comments as you had asked.